### PR TITLE
feat: deltalake `MERGE` support + additional features and bug fixes

### PR DESCRIPTION
**Pull request by ion-elgreco**
## New features
- Adds `MERGE` operation with couple flavors of it
- Adds get_output_stats to collect stats, can be extended per TypeHandler
- Adds support for reading array of partitions from upstream partitioned dataset

## Bug fixes
- Fixes config defaults, they weren't set before (made static type checker fail)
- Load input from upstream partitioned dataset would incorrectly always use start_dt, while if the asset that is being materialized is non-partitioned it should be the range

### CODE REVIEW

**Comment by ion-elgreco** [comment]
@PedramNavid @roeap can any of you have a look at this PR? :)

**Comment by ion-elgreco** [comment]
@benpankow @gibsondan @alangenfeld can someone review this PR?

File: python_modules/dagster/dagster/_core/storage/db_io_manager.py [hunk 1]
```diff
@@ -185,7 +185,7 @@ def _get_table_slice(
             # schema order of precedence: metadata, I/O manager 'schema' config, key_prefix
             if output_context_metadata.get("schema"):
                 schema = cast(str, output_context_metadata["schema"])
-            elif self._schema:
+            elif self._schema is not None:
```
Comment (PedramNavid) [inline python_modules/dagster/dagster/_core/storage/db_io_manager.py:188]: As discussed, this PR should limit changes only to the deltalake integration.

**Comment by PedramNavid** [comment]
hi @ion-elgreco, this is a community-supported integration, so the best person to review these changes would be @roeap when he has time.

**Pull request closed without merging**